standards/core/rln-contract.md

```diff
@@ -0,0 +1,301 @@
+- the size of the deposit MUST depend on the requested rate limit;
+- in case of a successful registration:
+	- the new membership MUST become _Active_;
+	- the current total rate limit MUST be incremented by the rate limit of the new membership;
+- a newly created membership MUST have an expiration time `T` and a grace period `G`.
+
+[^2]: A user-facing application SHOULD suggest default values for rate limits for the user.
+
+### Extend a membership
+
+Extending a membership is subject to the following conditions:
+- extension MUST fail if the membership is in any state other than _GracePeriod_;
+- the membership holder MUST be able to extend their membership;
+- any user except the membership holder MUST NOT be able to extend a membership;
+- after a successful extension, the membership MUST become _Active_.
+
+Holding a membership means controlling the private key from which the RLN commitment ID (i.e., public key) was derived.
+
+### Withdraw the deposit
+
+Deposit withdrawal is subject to the following conditions:
+- the membership holder MUST be able to withdraw their deposit;
+- any user except the membership holder MUST NOT be able to withdraw its deposit;
+- a deposit MUST be withdrawn in full;
+- a withdrawal MUST fail if the membership is not in _GracePeriod_, _Expired_, or _ErasedAwaitsWithdrawal_;
+- a membership MUST become _Erased_ after withdrawal.
+
+## Governance and upgradability
+
+At initial mainnet deployment, the contract MUST have an _Owner_.
+The _Owner_ MUST be able to change the values of all contract parameters.
+The _Owner_ MUST be able to pause any of the following contract functionalities:
+- register a membership;
+- extend a membership;
+- withdraw a deposit.
+
+At some point, the _Owner_ SHOULD renounce their privileges, and the contract MUST become immutable.
+Further upgrades, if necessary, SHOULD be done by deploying a new contract and migrating the membership set.
+
+## Implementation Suggestions
+
+User-facing application SHOULD suggest a few rate limits (tiers) to simplify their users' choice.
+The RECOMMENDED rate limits in a three-tier model are as follows:
+- `20` messages per epoch as low-tier;
+- `200` messages per epoch as mid-tier;
+- `600` messages per epoch as high-tier.
+Users SHOULD be able to select a custom rate limit under advanced settings.
+
+The RECOMMENDED pricing parameters are:
+
+| Parameter          | Value  |
+| ------------------ | ------ |
+| Accepted tokens    | `DAI`  |
+| Reference currency | `USD`  |
+| Pricing function   | linear |
+
+User-facing applications SHOULD save membership expiration dates in a local keystore during registration,
+and notify the user when their membership is about to expire.
+
+## Q&A
+
+### Why can't I withdraw a deposit from an _Active_ membership?
+
+The rationale for this limitation is to prevent an undesirable usage pattern where users make deposits and withdrawals in short succession.
```

Comment:
```suggestion
The rationale for this limitation is to prevent an undesirable usage pattern where users make deposits and withdrawals in short succession.
Such pattern may lead to network instability and should be carefully studied if seen as desirable.
```